core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -1055,12 +1056,36 @@ private TrivialRule() {
       super(any(MutableRel.class), any(MutableRel.class), 0);
     }
 
-    @Override public @Nullable UnifyResult apply(UnifyRuleCall call) {
+    @Override
+    public @Nullable UnifyResult apply(UnifyRuleCall call) {
       if (call.query.equals(call.target)) {
         return call.result(call.target);
       }
+      if (isSpa(call.target) && MutableRels.descendants(call.query)
+          .containsAll(MutableRels.descendants(call.target))) {
+        return call.result(call.query);
+      }
       return null;
     }
+
+    private boolean isSpa(MutableRel rel) {
```

Comment:
What does isSpa do? I'm not sure what Spa is in this context so it would help to have a more descriptive name.